modules/codegen/src/main/scala/com/twilio/guardrail/ProtocolGenerator.scala

```diff
@@ -142,48 +243,121 @@ object ProtocolGenerator {
     Free.pure(RandomType(clsName, tpe))
   }
 
-  def fromArray[F[_]](clsName: String, arr: ArrayModel, concreteTypes: List[PropMeta])(implicit R: ArrayProtocolTerms[F],
-                                                                                       A: AliasProtocolTerms[F]): Free[F, ProtocolElems] = {
+  def fromArray[F[_]](clsName: String, arr: ArrayModel, concreteTypes: List[PropMeta])(
+      implicit R: ArrayProtocolTerms[F],
+      A: AliasProtocolTerms[F]
+  ): Free[F, ProtocolElems] = {
     import R._
     for {
       tpe <- extractArrayType(arr, concreteTypes)
       ret <- typeAlias(clsName, tpe)
     } yield ret
   }
 
-  def fromSwagger[F[_]](swagger: Swagger)(implicit E: EnumProtocolTerms[F],
-                                          M: ModelProtocolTerms[F],
-                                          A: AliasProtocolTerms[F],
-                                          R: ArrayProtocolTerms[F],
-                                          S: ProtocolSupportTerms[F],
-                                          F: FrameworkTerms[F]): Free[F, ProtocolDefinitions] = {
+  case class ClassHierarchy(parentName: String, parentModel: Model, children: List[ClassHierarchy], discriminator: Option[String] = None)
+
+  /**
+    * returns objects grouped into hierarchies
+    */
+  def groupHierarchies(definitions: List[(String, Model)]): List[ClassHierarchy] = {
+
+    def firstInHierarchy(model: Model): Option[ModelImpl] =
+      (model match {
+        case elem: ComposedModel =>
+          definitions.collectFirst {
+            case (clsName, element) if elem.getInterfaces.asScala.headOption.exists(_.getSimpleRef == clsName) => element
+          }
+        case _ => None
+      }) match {
+        case Some(x: ComposedModel) => firstInHierarchy(x)
+        case Some(x: ModelImpl)     => Some(x)
+        case _                      => None
+      }
+
+    def children(cls: String, model: Model): List[ClassHierarchy] = definitions.collect {
+      case (clsName, comp: ComposedModel) if comp.getInterfaces.asScala.headOption.exists(_.getSimpleRef == cls) =>
+        classHierarchy(clsName, comp)
+    }
+
+    def classHierarchy(cls: String, model: Model): ClassHierarchy = ClassHierarchy(
+      cls,
+      model,
+      children(cls, model),
+      model match {
+        case m: ModelImpl     => Option(m.getDiscriminator)
+        case c: ComposedModel => firstInHierarchy(c).map(_.getDiscriminator)
+        case _                => None
```

Comment:
What is the use of a non-model as the child of a hierarchy? Would it be better for `classHierarchy` to return an `Option[ClassHierarchy]` or even a `Target.error` and make `discriminator` mandatory to avoid the `discriminator.get` down below?

Reply:
this comment hasn't been resolved

Reply:
this comment has been resolved. look at the code more attentively :)

Reply:
I hadn't looked at the code yet, as I'm not at my computer; just going off what comments hadn't been resolved by changes. I'll look.

Reply:
I can't see what this initially referred to, resolving this conversation